pepper/apps/trade_execution.c

```diff
@@ -0,0 +1,86 @@
+#include "trade_execution.h"
+#include "dex_common.h"
+#include "parsing.h"
+#include "hashing.h"
+
+void compute(struct In *input, struct Out *output) {
+    struct Private pInput = readPrivateInput();
+    // We need a way of faling an assert (e.g. assert_zero(1)).
+    // Hardcoding the one in code doesn't compile, so we have to pass and verify it.
+    assert_zero(input->one - 1);
+
+    // Assert that private input matches public
+    // Step 1: Balances (PedersenHash)
+    field254 balanceHashPedersen = hashPedersen(pInput.balances, 0, ACCOUNTS*TOKENS*BITS_PER_DECIMAL, BITS_PER_DECIMAL);
+    assert_zero(input->state - balanceHashPedersen);
+    // Step 2: Orders (PedersenHash)
+    field254 orderHashPedersen = hashPedersen(pInput.orders, 0, ORDERS*BITS_PER_ORDER, BITS_PER_ORDER);
+    assert_zero(input->orderHash - orderHashPedersen);
+    
+    // Step 3: Prices/Volumes (SHA Hash)
+    struct ShaResult hashBatchInfo = hashSHA(pInput.pricesAndVolumes, 0, (TOKENS*BITS_PER_DECIMAL)+(ORDERS*BITS_PER_DECIMAL), 256);
+    assert_zero(input->hashBatchInfo[0] - hashBatchInfo.left);
+    assert_zero(input->hashBatchInfo[1] - hashBatchInfo.right);
+
+    // Parse all private input
+    struct Order orders[ORDERS] = {0};
+    parseOrders(pInput.orders, 0, orders);
+    struct Balance balances[ACCOUNTS] = {0};
+    parseBalances(pInput.balances, 0, balances);
+    field254 prices[TOKENS] = {0};
+    parsePrices(pInput.pricesAndVolumes, 0, prices);
+    struct Volume volumes[ORDERS] = {0};
+    parseVolumes(pInput.pricesAndVolumes, (TOKENS*100), volumes);
+
+    // Accumulator variables
+    field254 buyVolumes[TOKENS]= {0};
+    field254 sellVolumes[TOKENS]= {0};
+    field254 totalSurplus = 0;
+
+    // validate each order and update balances
+    uint32_t index;
```

Comment:
What is uint32_t? Is this our own type or a c primitive?

Reply:
Standard c type 32bit wide unsigned integers (https://en.cppreference.com/w/cpp/types/integer)

I use it whenever we have something that is not a field element conceptually (e.g. an index)